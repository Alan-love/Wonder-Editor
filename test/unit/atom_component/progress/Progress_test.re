open Wonder_jest;

open Expect;

open Expect.Operators;

open Sinon;

let _ =
  describe("Progress component", () => {
    let sandbox = getSandboxDefaultVal();

    let _buildProgressComponent = () =>
      ReactTestRenderer.create(<Progress />);

    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());
    });
    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test progress component", () => {
      let _buildFakeProgressDom = (~display="flex", ()) => {
        "style": {
          "display": display,
        },
      };

      test("test snapshot", () =>
        _buildProgressComponent() |> ReactTestTool.createSnapshotAndMatch
      );

      describe("operate progress visible", () => {
        let progressDom = ref(Obj.magic(-1));

        beforeEach(() => {
          _buildProgressComponent();
          progressDom := _buildFakeProgressDom(~display="flex", ());
          DomTool.stubFakeDomForQuerySelector(
            sandbox,
            "wonder-progress",
            progressDom^,
          );
        });

        describe("show progress", () =>
          test("show progress dom", () => {
            ProgressUtils.hide |> StateLogicService.getAndSetEngineState;
            ProgressUtils.show |> StateLogicService.getAndSetEngineState;

            (progressDom^)##style##display |> expect == "flex";
          })
        );

        describe("hide progress", () =>
          test("hide progress dom", () => {
            ProgressUtils.hide |> StateLogicService.getAndSetEngineState;

            (progressDom^)##style##display |> expect == "none";
          })
        );

        test("if percent === 100, hide progress", () => {
          ProgressUtils.changePercent(100)
          |> StateLogicService.getAndSetEngineState;

          (progressDom^)##style##display |> expect == "none";
        });
        test("if finish, hide progress", () => {
          ProgressUtils.finish |> StateLogicService.getAndSetEngineState;

          (progressDom^)##style##display |> expect == "none";
        });
      });

      describe("test off custom global event in willUnmount", () => {
        let progressDom = ref(Obj.magic(-1));

        beforeEach(() => {
          _buildProgressComponent();
          progressDom := _buildFakeProgressDom(~display="flex", ());
          DomTool.stubFakeDomForQuerySelector(
            sandbox,
            "wonder-progress",
            progressDom^,
          );
        });

        test("test off hide event", () => {
          _buildProgressComponent();

          Progress.Method.willUnmount();

          ProgressUtils.hide |> StateLogicService.getAndSetEngineState;

          (progressDom^)##style##display |> expect == "flex";
        });
      });
    });
  });