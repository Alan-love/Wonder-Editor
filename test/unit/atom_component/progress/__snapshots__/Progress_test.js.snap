// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress component test progress component test snapshot 1`] = `
<article
  className="wonder-progress"
>
  <div
    className="progress-content"
  >
    <div
      className="content-percent"
      style={
        Object {
          "width": "50%",
        }
      }
    >
      50%
    </div>
  </div>
  <div
    className="progress-bg"
  />
</article>
`;
